### PR TITLE
NEW Auto patch releases

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripeltd/product-issues/issues/754

Example CI runs in issue description

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -712,3 +725,166 @@ jobs:
...
+    env:
+      GITHUB_REPOSITORY: ${{ github.repository }}
+      GITHUB_REF_NAME: ${{ github.ref_name }}
+      GITHUB_SHA: ${{ github.sha }}
+      LATEST_LOCAL_SHA: ${{ needs.tests.outputs.latest_local_sha }}
+    outputs:
+      do_release: ${{ steps.determine-if.outputs.do_release }}
+      release_notes: ${{ steps.determine-if.outputs.release_notes }}
+      latest_tag: ${{ steps.determine-if.outputs.latest_tag }}
+      next_tag: ${{ steps.determine-if.outputs.next_tag }}
```
Comment (ChloeSartorelli) [inline .github/workflows/ci.yml]: Looks like we're only using two of these outputs in `patchrelease` - what are the other two for?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -712,3 +725,166 @@ jobs:
         with:
           name: ${{ env.artifacts_name }}
           path: artifacts
+
+  gaugerelease:
+    name: Check unreleased changes
+    runs-on: ubuntu-latest
+    needs: tests
+    # TODO: restricted to silverstripe/silverstripe-framework - remove once we've validated this whole thing works
+    if: ${{ github.repository == 'silverstripe/silverstripe-framework' && github.repository_owner == 'silverstripe' && (github.event_name == 'push' ||  github.event_name == 'schedule') }}
```
Comment (ChloeSartorelli) [inline .github/workflows/ci.yml:734]: Please also create an issue for this TODO if you haven't already.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -712,3 +725,166 @@ jobs:
...
+      - name: Determine if should patch release
+        id: determine-if
+        run: |
+          DO_RELEASE=1
+
+          # Must be on a minor branch to do a patch release
+          if ! [[ $GITHUB_REF_NAME =~ ^[0-9]+\.[0.9]+$ ]]; then
+            echo "Not patch releasing because not on a minor branch"
+            DO_RELEASE=0
+          fi
```
Comment (ChloeSartorelli) [inline .github/workflows/ci.yml:754]: We should set our output and `exit 0` here if we've already determined we don't want to do a release. Same with anywhere below where we're setting it to 0.

Given that happens a few times it probably makes sense to set that up as a function and just call the function whenever we determine we don't wanna do a release.

e.g:
```suggestion
          no_release () {
            echo "do_release output is 0";
            echo "do_release=0" >> $GITHUB_OUTPUT;
            exit 0;
          }

          # Must be on a minor branch to do a patch release
          if ! [[ $GITHUB_REF_NAME =~ ^[0-9]+\.[0.9]+$ ]]; then
            echo "Not patch releasing because not on a minor branch"
            no_release
          fi
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -712,3 +725,166 @@ jobs:
...
+
+          # Validate that this commit is that latest commit for the branch using GitHub API
+          # We need to check this in case re-rerunning an old job and there have been new commits since
+          if [[ $DO_RELEASE == "1" ]]; then
```
Comment (ChloeSartorelli) [inline .github/workflows/ci.yml:758]: Given the exit above, we don't need this condition anymore.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -712,3 +725,166 @@ jobs:
...
+              DO_RELEASE=0
+            fi
+          fi
+
+          # Check is there is an existing tag on the branch using GitHub API
+          # Note cannot use local `git tag` because actions/checkout by default will not checkout tags
+          # and you need to checkout full history in order to get them
+          LATEST_TAG=""
+          NEXT_TAG=""
+          if [[ $DO_RELEASE == "1" ]]; then
```
Comment (ChloeSartorelli) [inline .github/workflows/ci.yml:789]: Given the exit above, we don't need this condition anymore.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -712,3 +725,166 @@ jobs:
...
+              NEXT_TAG="$MAJOR.$MINOR.$((PATCH+1))"
+              echo "NEXT_TAG is $NEXT_TAG"
+              echo "next_tag=$NEXT_TAG" >> $GITHUB_OUTPUT
+            fi
+          fi
+
+          # Check if there is anything relevant to release using GitHub API
+          # The API results include all merged pull-requests, not commits
+          # Pull-requests prefixed with MNT or DOC will not be considered relevant for releasing
+          if [[ $DO_RELEASE == "1" ]]; then
```
Comment (ChloeSartorelli) [inline .github/workflows/ci.yml:821]: Given the exit above, we don't need this condition anymore.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -712,3 +725,166 @@ jobs:
...
+                fi
+              fi
+            done < __body.json
+            if [[ $HAS_THINGS_TO_RELEASE == "0" ]]; then
+              echo "Not patch releasing because there is nothing relevant to release"
+              DO_RELEASE=0
+            fi
+          fi
+          echo "do_release output is $DO_RELEASE"
+          echo "do_release=$DO_RELEASE" >> $GITHUB_OUTPUT
```
Comment (ChloeSartorelli) [inline .github/workflows/ci.yml:851]:
```suggestion
          echo "do_release output is 1"
          echo "do_release=1" >> $GITHUB_OUTPUT
```

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -712,3 +725,166 @@ jobs:
...
+          if ! [[ $GITHUB_REF_NAME =~ ^[0-9]+\.[0.9]+$ ]]; then
+            echo "Not patch releasing because not on a minor branch"
+            DO_RELEASE=0
+          fi
+
+          # Validate that this commit is that latest commit for the branch using GitHub API
+          # We need to check this in case re-rerunning an old job and there have been new commits since
+          if [[ $DO_RELEASE == "1" ]]; then
+            # https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28
+            AMP='&'
```
Comment (ChloeSartorelli) [inline .github/workflows/ci.yml]: Why is this needed instead of just putting `&` directly in the URL?

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -712,3 +725,166 @@ jobs:
...
+            fi
+          fi
+
+          # Check is there is an existing tag on the branch using GitHub API
+          # Note cannot use local `git tag` because actions/checkout by default will not checkout tags
+          # and you need to checkout full history in order to get them
+          LATEST_TAG=""
+          NEXT_TAG=""
+          if [[ $DO_RELEASE == "1" ]]; then
+            # https://docs.github.com/en/rest/git/tags?apiVersion=2022-11-28
```
Comment (ChloeSartorelli) [inline .github/workflows/ci.yml]:
```suggestion
            # https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#list-repository-tags
```

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -712,3 +725,166 @@ jobs:
...
+            RESP_CODE=$(curl -w %{http_code} -s -o __response.json \
+            -X GET https://api.github.com/repos/${GITHUB_REPOSITORY}/tags?per_page=100 \
+            -H "Accept: application/vnd.github+json" \
+            -H "Authorization: Bearer ${{ github.token }}" \
+            -H "X-GitHub-Api-Version: 2022-11-28" \
+            )
+            if [[ $RESP_CODE != "200" ]]; then
+              echo "Unable to read list of commits - HTTP response code was $RESP_CODE"
```
Comment (ChloeSartorelli) [inline .github/workflows/ci.yml]:
```suggestion
              echo "Unable to read list of tags - HTTP response code was $RESP_CODE"
```

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -712,3 +725,166 @@ jobs:
...
+          LATEST_TAG=""
+          NEXT_TAG=""
+          if [[ $DO_RELEASE == "1" ]]; then
+            # https://docs.github.com/en/rest/git/tags?apiVersion=2022-11-28
+            RESP_CODE=$(curl -w %{http_code} -s -o __response.json \
+            -X GET https://api.github.com/repos/${GITHUB_REPOSITORY}/tags?per_page=100 \
+            -H "Accept: application/vnd.github+json" \
+            -H "Authorization: Bearer ${{ github.token }}" \
+            -H "X-GitHub-Api-Version: 2022-11-28" \
+            )
```
Comment (ChloeSartorelli) [inline .github/workflows/ci.yml]: We can use https://docs.github.com/en/rest/git/refs?apiVersion=2022-11-28#list-matching-references instead

e.g. where `GITHUB_REPOSITORY` is "silverstripe/silverstripe-framework" and the branch being pushed to is "4.13":
https://api.github.com/repos/silverstripe/silverstripe-framework/git/matching-refs/tags/4.13 gives us all tags that start with 4.13! We just need to grab the list item in the list, which simplifies the code below that's trying to get the latest tag.

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -712,3 +725,166 @@ jobs:
...
+              DO_RELEASE=0
+            else
+              MAJOR=${BASH_REMATCH[1]}
+              MINOR=${BASH_REMATCH[2]}
+              PATCH=${BASH_REMATCH[3]}
+              NEXT_TAG="$MAJOR.$MINOR.$((PATCH+1))"
+              echo "NEXT_TAG is $NEXT_TAG"
+              echo "next_tag=$NEXT_TAG" >> $GITHUB_OUTPUT
+            fi
+          fi
```
Comment (ChloeSartorelli) [inline .github/workflows/ci.yml]: I'm not reviewing this part of the code right now, as I think it will change significantly when we swap to fetching tags with the matching-refs API per https://github.com/silverstripe/gha-ci/pull/72#discussion_r1247263493

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -712,3 +725,166 @@ jobs:
...
+                if [[ "$line" =~ ^\*[[:space:]](.+?)$ ]]; then
+                  PR_TITLE=${BASH_REMATCH[1]}
+                  if ! [[ $PR_TITLE =~ ^(MNT|DOC) ]]; then
+                    HAS_THINGS_TO_RELEASE=1
+                  fi
+                else
+                  IS_PRS_SECTION=0
+                fi
+              fi
+            done < __body.json
```
Comment (ChloeSartorelli) [inline .github/workflows/ci.yml]: Using release notes means we're only looking at the names of PRs - which frankly are irrelevant. Consider for example a PR named "fix broken builds" where the commit message is "MNT blah blah broken build". The release notes will have "fix broken builds" so this will say "yay that's a fix!" But the commit is MNT.

Use https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#compare-two-commits instead. This gives us some nice json where we can check each commit's message. You can cleanly compare from the latest tag to the current commit sha. See for example:
https://api.github.com/repos/silverstripe/silverstripe-framework/compare/4.13.8...5303e08d34c80348514a6c1d90815450b2518921

This API also gives us more information (e.g. about the actual files changes) in case we wanted to do some more specific stuff with this later on.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 9 above)
Comment (ChloeSartorelli) [reply .github/workflows/ci.yml]: Superceded by https://github.com/silverstripe/gha-ci/pull/72#discussion_r1247263493

File: .github/workflows/ci.yml (hunk 2 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml:734]: Have created issue https://github.com/silverstripeltd/product-issues/issues/769

File: .github/workflows/ci.yml (hunk 1 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml]: They're redundant, have removed

File: .github/workflows/ci.yml (hunk 8 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml]: Because bash does weird things :-)  It was truncating the string from memory.  It needs to be escaped using some method, thing method happens to work

File: .github/workflows/ci.yml (hunk 11 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml]: I'd say it'll probably actually complicate the code below due to the need to strip the `refs/tags/` out of the `ref` attribute.  Currently we use
`LATEST_TAG=$(jq -r '.[].name' __response.json | grep -E ^$ESCAPED_REF_NAME\.[0-9]+$ | sort -V -r | head -n 1)`

We'd need to change it to something like

`LATEST_TAG=$(jq -r '.[].ref' __response.json | <remove refs/tags/> | grep -E ^[0-9]+\.[0-9]+\.[0-9]+$ | sort -V -r | head -n 1)`

Though I'm not sure how to do the `<remove refs/tags/>` bit in bash at this point in the pipe chain

I would prefer to do it this way just not needing to worry about the 100 tag limitation, though as mentioned offline the [tags API for framework](https://api.github.com/repos/silverstripe/silverstripe-framework/tags?per_page=100) goes all the way back to 4.5.1 which is way out of support, so the 100 limit should be fine in practice

File: .github/workflows/ci.yml (hunk 10 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml]: Updated

File: .github/workflows/ci.yml (hunk 12 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml]: We still need this code no matter how we read tags.  We need to increment the latest semver tag by 1 patch

File: .github/workflows/ci.yml (hunk 3 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml:754]: I don't like this recommended change, it does not match the coding style we've used in any other github action - we do not use bash functions in any of them.  I much prefer just using a variable I think it's more readable and easier to follow.

File: .github/workflows/ci.yml (hunk 6 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml:821]: I'd prefer to keep the variable rather than switch to a function

File: .github/workflows/ci.yml (hunk 5 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml:789]: I'd prefer to keep the variable rather than switch to a function

File: .github/workflows/ci.yml (hunk 4 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml:758]: I'd prefer to keep the variable rather than switch to a function

File: .github/workflows/ci.yml (hunk 7 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml:851]: This doesn't make sense since I'd like to keep the `$DO_RELEASE` variable

File: .github/workflows/ci.yml (hunk 13 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml]: Have updated to use commits - have tested that it works correctly with a [MNT](https://github.com/emteknetnz/silverstripe-config/actions/runs/5420959807) followed by a [ENH](https://github.com/emteknetnz/silverstripe-config/actions/runs/5420978723) commit (didn't use PR's for these hence [auto release notes](https://github.com/emteknetnz/silverstripe-config/releases/tag/2.0.4) don't mention them)

File: .github/workflows/ci.yml (hunk 11 above)
Comment (ChloeSartorelli) [reply .github/workflows/ci.yml]: We know that we only have _relevant_ tags in this list. We also know that they're in order from either oldest to newest or lowest semver to highest semver (which for our purposes are identical sort orders). So we can just grab the last item in the list without resorting or using grep etc - that would look something like this:
```bash
LATEST_TAG=$(jq -r '.[-1:][].ref' __response.json)
```
We can then do something like `LATEST_TAG=${LATEST_TAG#refs/tags/}` to remove `refs/tags` - but I'm sure there are plenty of other ways to do it once we have it as a variable.

If you do want to keep the grep and sort in, just change your regex to `^ref/tags/[0-9]+\.[0-9]+\.[0-9]+$` and remove the `ref/tags/` part after we've got it in a variable. So:
```bash
LATEST_TAG=$(jq -r '.[].ref' __response.json | grep -E ^refs\/tags\/[0-9]+\.[0-9]+\.[0-9]+$ | sort -V -r | head -n 1)
LATEST_TAG=${LATEST_TAG#refs/tags/}
```

If you want to do everything in one line like you have been, you can use a positive lookbehind to remove the `refs/tags/` part as part of the match, like so:
```bash
LATEST_TAG=$(jq -r '.[].ref' __response.json | grep -Po '(?<=^refs\/tags\/)[0-9]+\.[0-9]+\.[0-9]+$' | sort -V -r | head -n 1)
```

File: .github/workflows/ci.yml (hunk 12 above)
Comment (ChloeSartorelli) [reply .github/workflows/ci.yml]: Fair enough, I hadn't read through this and assumed it was _all_ part of the json parsing but now that I look a little through it I see that's not the case.

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -712,3 +725,144 @@ jobs:
...
+              echo "Unable to read list of tags - HTTP response code was $RESP_CODE"
+              exit 1
+            fi
+            # Get the latest tag
+            # Escape for regex e.g. 2.0 => 2\.0
+            ESCAPED_REF_NAME="${GITHUB_REF_NAME/./"\."}"
+            LATEST_TAG=$(jq -r '.[].name' __response.json | grep -E ^$ESCAPED_REF_NAME\.[0-9]+$ | sort -V -r | head -n 1)
+            echo "LATEST_TAG is $LATEST_TAG"
+            echo "latest_tag=$LATEST_TAG" >> $GITHUB_OUTPUT
+            if ! [[ $LATEST_TAG =~ ([0-9]+)\.([0-9]+)\.([0-9]+) ]]; then
```
Comment (ChloeSartorelli) [inline .github/workflows/ci.yml:805]: I don't see the point of this condition, since we're already using REGEX to get the latest tag name anyway? Why don't we just require the latest tag to be a patch tag in the first set of regex?

File: .github/workflows/ci.yml (hunk 3 above)
Comment (ChloeSartorelli) [reply .github/workflows/ci.yml:754]: I don't think not having used a function before is a good reason to not use one here if the function improves this code - which I think it does. To my eye, the function is a lot more readable and easier to maintain.
But we're not going to see eye to eye on that and since it's subjective I'm okay leaving it the way you have it.

File: .github/workflows/ci.yml [hunk 15]
```diff
@@ -712,3 +725,144 @@ jobs:
...
+            DO_RELEASE=0
+          fi
+
+          # Validate that this commit is that latest commit for the branch using GitHub API
+          # We need to check this in case re-rerunning an old job and there have been new commits since
+          if [[ $DO_RELEASE == "1" ]]; then
+            # https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28
+            AMP='&'
+            RESP_CODE=$(curl -w %{http_code} -s -o __response.json \
+            -X GET https://api.github.com/repos/${GITHUB_REPOSITORY}/commits?sha=${GITHUB_REF_NAME}${AMP}per_page=1 \
```
Comment (ChloeSartorelli) [inline .github/workflows/ci.yml]:
```suggestion
            RESP_CODE=$(curl -w %{http_code} -s -o __response.json \
            -X GET "https://api.github.com/repos/${GITHUB_REPOSITORY}/commits?sha=${GITHUB_REF_NAME}per_page=1" \
```
as per https://github.com/silverstripe/gha-ci/pull/72#discussion_r1249926926

File: .github/workflows/ci.yml (hunk 8 above)
Comment (ChloeSartorelli) [reply .github/workflows/ci.yml]: Please just use normal string escaping instead. I'll add that as a separate suggestion.

File: .github/workflows/ci.yml (hunk 11 above)
Comment (ChloeSartorelli) [reply .github/workflows/ci.yml]: I think knowing we _are only_ dealing with tags for the minor release we care about is a big plus, and not having the 100 tag limit is a theoretical win as well. I've given three ways to handle this, ranging from "just grab the last item 'cause we know it's the one we need" to "keep all your existing logic and just change the regex slightly to account for the prefix".

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 14 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml:805]: $LATEST_TAG could be blank (or something else, I don't know what it would be returned) if there are zero tags that match

File: .github/workflows/ci.yml (hunk 15 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml]: Updated

File: .github/workflows/ci.yml (hunk 8 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml]: I did try it initially, I literally don't know what I'm supposed to put in here.  Bash is super fussy.  While this isn't great it's good enough?

File: .github/workflows/ci.yml (hunk 15 above)
Comment (ChloeSartorelli) [reply .github/workflows/ci.yml]: I did test this before I suggested it. It does work. I don't see what's the problem?

File: .github/workflows/ci.yml (hunk 8 above)
Comment (ChloeSartorelli) [reply .github/workflows/ci.yml]: Resolving - discussion can be on the part with the actual code suggestion.

Comment (emteknetnz) [reply .github/workflows/ci.yml]: My bad sorry, missed that, updated

File: .github/workflows/ci.yml (hunk 15 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml]: Updated

File: .github/workflows/ci.yml (hunk 11 above)
Comment (emteknetnz) [reply .github/workflows/ci.yml]: Updated to use the positive lookbehind within the pipe chain, works nice cheers

**Review by GuySartorelli — APPROVED** [review]
